### PR TITLE
ARROW-15116: [Python] Expose invalid_row_handler for CSV reader

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good, added a few minor comments.

A bit orthogonal to this PR, but while looking at it: it could be an extension to allow an additional handler return value of "fill_null" (or another name), which could signal that in case the row has fewer fields, the missing fields can be filled with nulls (this is actually the default behaviour in pandas). Of course that could also be its own separate option, but letting users handle it through this functionality would avoid adding yet another option.

File: python/pyarrow/tests/test_csv.py [hunk 1]
```diff
@@ -109,6 +109,27 @@ def check_options_class_pickling(cls, **attr_values):
...
 
+class InvalidRowHandler:
+    def __init__(self, result):
+        self.result = result
+        self.rows = []
+
+    def __call__(self, row):
+        self.rows.append(row)
+        if isinstance(self.result, Exception):
+            raise self.result
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/tests/test_csv.py:120]: I hoped to use it, but the exception is simply printed out and pytest has no simple way of catching that. I'll just remove this instead.

File: python/pyarrow/_csv.pyx [hunk 2]
```diff
@@ -254,15 +275,21 @@ cdef class ParseOptions(_Weakrefable):
         Whether empty lines are ignored in CSV input.
         If False, an empty line is interpreted as containing a single empty
         value (assuming a one-column CSV file).
+    invalid_row_handler : callable, optional (default None)
+        If not None, this object is called for each CSV row that fails
+        parsing (because of a mismatching number of columns).
+        It should accept a single InvalidRow argument and return either
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/_csv.pyx:281]: I'll add a class docstring with attribute descriptions.

[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/_csv.pyx:281]: Note that numpydoc's support for namedtuple is rather poor:
https://github.com/numpy/numpydoc/issues/257

**Comment by pitrou** [comment]
> A bit orthogonal to this PR, but while looking at it: it could be an extension to allow an additional handler return value of "fill_null" (or another name), which could signal that in case the row has fewer fields, the missing fields can be filled with nulls (this is actually the default behaviour in pandas).

That's at least theoretically doable, perhaps you can open a separate JIRA for that?

**Comment by pitrou** [comment]
Cool, now I have to document everything twice. Thank you numpydoc...
```
pyarrow._csv.InvalidRow
PR01: Parameters {'text', 'expected_columns', 'number', 'actual_columns'} not documented
```

**Comment by pitrou** [comment]
For some reason AppVeyor is broken on the Apache account. It works on my fork: https://ci.appveyor.com/project/pitrou/arrow/builds/42033790

**Pull request closed without merging**